📖add quota validation plugin proposal (#244)
kubernetes-sigs/cluster-api-provider-nested: docs/proposals/20220206-quota.md

## Motivation

* Per namespace quota admission control is not suitable for multi-tenant virtual cluster environment. This is because It requires tenant to coordinate with super cluster admin to create namespaces and setup quotas in advance. It also forces super cluster admin to map namespaces with tenants in order to control overall quota usage for each tenant.
* Cross namespace and per tenant overall quota admission control provides a true virtual cluster experience for tenant. It allows tenant freely creating/updating/deleting namespaces with guarantee of not breaking  resource quota allowance in super cluster. It also reduce resource fragmentation in tenant space. 
* Cross namespace quota admission control needs a centralized engine to validate resource creation. Traditional approach is to use validation Webhook. However, using Webhook in multi-tenant cluster environment has following issues:
    * for a Webhook installed in tenant virtual cluster: as tenants have admin privileges in their virtual cluster, the quota validation Webhook can be deleted by tenant and is not a truly enforced mechanism.

>     * for a Webhook installed in super cluster: it becomes a bottle neck for all tenants and impacts multiple threading model implemented in pod syncer.  In addition the Webhook needs to map namespaces with tenants, and other information originally kept in syncer to performance validation, which is not efficient.

Reviewer: ```suggestion
    * for a Webhook installed in super cluster: it becomes a bottle neck for all tenants and impacts multiple threading model implemented in pod syncer.  
```

They can always get this data from there object labels.

* Given quota admission control is a proprietary implementation for different cloud environments. Providing an universal solution is not pratical. This is why in this document we propose to build a plugin to let users to easily integrate their quota validation logics into pod-syncer.

### Goals

> * Propose a quota validation plugin for user to integrate proprietary quota management functions into pod-syncer.

Reviewer: ```suggestion
* Propose a quota validation plugin for user to integrate custom quota management functions into pod-syncer.
```

* Define per tenant locking mechanism in pod-syncer to preserve parallel processing of tenants’ requests and prevent race condition while checking quota usage.
* Give an example of quota validation plugin registration and initialization.

### Non-Goals

* Queuing mechanism used for managing pods that exceeds quota limit.
* Define design and implementations for quota provision, quota validation and quota usage tracking.

## Proposal

### User Stories and Requirements

* Enforce tenant overall quota usage in super cluster but allow tenant freely creating namespace, and managing separate quota in virtual cluster.
* Enable cross namespace quota admission control in super cluster to reduce resource fragementation.
* Allow cloud admin to define own quota validation implementation and easy integration into pod-syncer

> * Handle tenant’s resource requests in parallel and minimize interferences of large batch requests from noisy tenant.

Reviewer: Can we write these more like:

**As a _____** _(cluster admin/operator/etc)_ **I want ________** _(feature eg 'be provide a custom quota handler that stops syncing')_ **so that _____** _(reason, eg 'I can customize what pods sync to the super cluster')_

* pod-syncer has all knowledge of namespace/tenant mapping, resource usage status and resource request details.
* All kind resource requests, e.g. through job creation, deployment, or from replica-set, all end up being processed in pod-syncer, which reduce the capturing and converting different resource requests.
* pod-syncer has all privilege of controlling resources in super cluster, able to hold tenant’s pod in pending state, and re-queue the pending pods periodically.
* the syncer has unique queuing mechanism that allows locking of one tenant in a short period of time without interfering other tenants.

Following diagram depicts how the validation plugin works together with pod-syncer.
![Resource Admission Control Plugin](images/quota/plugin.png)
When pod syncer reconciler receives pod creation request:

* it first checks if quota validation flag is on, if not, it goes ahead with normal pod creation procedure.
* if quota validation flag is on, pod-syncer searches for quota validation plugin that is dynamically registered at syncer starting time, and call the quota validation plugin

> * the pod-syncer first locks tenant, then get current tenant quota usage, compares available quota with requested quota from pod, if there is enough quota, pod syncer will creates the pod in super cluster and removes the per tenant lock.

Reviewer: We should explain by how this will be implemented upstream, not by how a validator should be written since that could be anything.

```suggestion
* the pod-syncer first locks tenant, then calls the quota validation plugin and if the validation is successful the pod syncer will create the pod in super cluster and removes the per tenant lock.
```

> * if there no enough quota, the per tenant lock will be released. The creation reconcile routine can return with no-requeue option, so the pod will not be putting back into reconcile queue immediately. The un-synced pod will be in pending state. The pod-syncer patroller will re-queue the pending pod later. 

Reviewer: ```suggestion
* if not successful, the per tenant lock will be released. The creation reconcile routine can return with no-requeue option, so the pod will not be putting back into reconcile queue immediately. The un-synced pod will be in pending state. The pod-syncer patroller will re-queue the pending pod later. 
```

**Validation Plugin**

> A ValidationPluginInterface will be defined to present all types of validation plugins. User proprietary quota validation plugin will be derived from it.

Reviewer: ```suggestion
A ValidationPluginInterface will be defined to present all types of validation plugins. Use custom quota validation plugin will be derived from it.
```

```
type Plugin struct {
   Registration *Registration // registration, as initialized   
   instance     interface{}
   err          error
}
```

```
type Registration struct {
   ID string   // ID of the plugin   
   
   // InitFn is called when initializing a plugin. The registration and context are passed in.   
   InitFn func(*mc.MultiClusterController) (interface{}, error)
   
   Disable bool  // Disable the plugin from loading   
}
```

The quota validation plugin is initialized when pod-syncer controller getting created in pkg/syncer/resource/pod/controller.go

```
func NewPodController(config *config.SyncerConfiguration, client clientset.Interface,   informer informers.SharedInformerFactory,   vcClient vcclient.Interface,   vcInformer vcinformers.VirtualClusterInformer,   options manager.ResourceSyncerOptions) (manager.ResourceSyncer, error) {
...
plugins := featureplugin.DefaultFeatureRegister.List()
for _, p := range plugins {
    if p.ID == "quotavalidationplugin" {
        pp, _ := p.Init(c.MultiClusterController).Instance()
        c.plugin = pp.(featureplugin.ValidationPluginInterface)
        break
    }
}
...
}
```

> Users need to register their proprietary quota validation logic in init function:

Reviewer: ```suggestion
Users need to register their custom quota validation logic in init function:
```

   Cond        *sync.Mutex
   }
```

Per tenant lock is invoked before the quota validation is called in “reconcilePodCreate” function of pod syncer:

```
func (c *controller) reconcilePodCreate(clusterName, targetNamespace, requestUID string, vPod *v1.Pod) error {
...
if c.plugin != nil {
            t := (c.plugin).GetTenantLocker(clusterName)
            if t == nil {
                return errors.NewBadRequest("cannot get tenant")
            }
            t.Cond.Lock()
            defer t.Cond.Unlock()
            if !(c.plugin).ValidationPlugin(&newObj, clusterName) {
                ...
        }
    }
...
pPod, err = c.client.Pods(targetNamespace).Create(context.TODO(), pPod, metav1.CreateOptions{})
...
}
```

The per tenant lock allows syncer to process requests in parallel for different tenants.

#### Tenant Map

A tenant map can be kept in quota validation plugin

```
type Quotaplugin struct {
   Name string     
   alltenants map[string]Tenant   
   *sync.RWMutex
}
```

A read lock will be used when getting tenant record from tenant map. A write lock is needed when tenant map gets updated periodically (every 2 hour or longer).

#### Resource Quota Object Syncing

> A resource quota object can be deployed to each namespace in super cluster. The resource qutoa object is used for counting resource usage for each priority class. The resource quota object is fully owned by syncer and does not need to be synced upward to virtual cluster.  

Reviewer: We can remove this section since it's specific to how we're aggregating quota.

```suggestion
```
